Add support for generating a consumer for CSS property that use group terms

weinig opened the pull request:
#### c652c23c52e249db4208beb501ce900a24fe0502
<pre>
Add support for generating a consumer for CSS property that use group terms
<a href="https://bugs.webkit.org/show_bug.cgi?id=289305">https://bugs.webkit.org/show_bug.cgi?id=289305</a>

Reviewed by Darin Adler.

Adds support for generating consumers with &quot;match one or more any order&quot; (eg `[ &lt;foo&gt; || &lt;bar&gt; ]`),
&quot;match all any order&quot; (eg `[ &lt;foo&gt; &amp;&amp; &lt;bar&gt; ]`), &quot;match all ordered&quot; (eg `[ &lt;foo&gt; &lt;bar&gt; ]`) and
optional terms (e.g. `&lt;foo&gt;?`).

A new annotation, @(preserves-order) was needed for properties with &quot;any order&quot; grammars (eg || or &amp;&amp;)
but where the order in which things are declared is important such as the `paint-order` property. In
some cases this has been used to maintain existing serialization behavior, such as hanging-punctuation,
but it is not clear that this is necessarily correct and should be evaluated separately.

Together, these allow generation of:
    font-variant-east-asian
    font-variant-ligatures
    font-variant-numeric
    text-autospace
    fill
    stroke
    speak-as
    text-indent
    text-transform
    contain
    hanging-punctuation
    text-decoration-line
    -webkit-text-decorations-in-effect
    text-emphasis-position
    text-emphasis-style
    touch-action
    @font-face font-variant-east-asian
    @font-face font-variant-ligatures
    @font-face font-variant-numeric

Additionally, this change adds generation for the following two properties/descriptors missed
in earlier passes:
    font-width
    @font-face font-weight

While adding generation for font-width, I noticed a few places were still calling
it font-stretch, so I updated those to match the new naming.

* LayoutTests/editing/execCommand/toggle-text-decorations-expected.txt:
* LayoutTests/editing/execCommand/toggle-text-decorations.html:
* LayoutTests/fast/css3-text/css3-text-decoration/getComputedStyle/getComputedStyle-text-decoration-shorthand-ordering-expected.txt:
* LayoutTests/fast/css3-text/css3-text-decoration/getComputedStyle/getComputedStyle-text-decoration-shorthand-ordering.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/parsing/text-decoration-line-valid-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/editing/run/strikethrough_2001-last-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/editing/run/underline_2001-last-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/strikethrough_2001-last-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/underline_2001-last-expected.txt:
    - The serialization of text-decorations-line has changed to match the grammar order which is what
      WPT expects.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Contain.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Contain.h: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PointerEvents.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PointerEvents.h: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Speech.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Speech.h: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Text.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Text.h: Removed.
    - Remove no longer needed files.

* Source/WebCore/css/scripts/process-css-properties.py:
    - Replaces single GroupTerm class with MatchAllOrderedTerm, MatchAllAnyOrderTerm
      and MatchOneOrMoreAnyOrderTerm classes.
    - Add support for &apos;type&apos; and &apos;no-single-item-opt&apos; annotations to the new term types.
    - Add term generators for the new term types.
    - Add pass-through term generation for OptionalTerm. Other generators currently
      just check if the generators instance matches to implement optional semantics.
    - Add additional comments to generated output with the grammar definitions.

* Source/WebCore/css/CSSProperties.json:
    - Update properties that can be generated to use the grammar.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
    - Call now generated functions instead of hand written ones.

* Source/WebCore/css/CSSFontFaceSet.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+SVG.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+SVG.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Text.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Text.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.h:
    - Remove now generated consume functions.
    - Call now generated functions instead of hand written ones.
    - Update uses of &apos;font-stretch&apos; naming to use &apos;font-width&apos; terminology.

* Source/WebCore/css/scripts/test/TestCSSProperties.json:
    - Add test cases for the different grouping types.

* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleDeclaration+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
    - Update test results.

Canonical link: <a href="https://commits.webkit.org/291871@main">https://commits.webkit.org/291871@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7e9cd45130034bc6d069f7fc47dc3b99527d542

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94248 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13835 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3601 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99260 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44775 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96298 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14135 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22265 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71899 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29239 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97250 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10494 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85101 "Found 1 new API test failure: TestWebKitAPI.WKWebExtensionAPIDevTools.MessagePassingFromPanelToBackground (failure)") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52245 "Found 1 new API test failure: /WPE/TestWebKitWebView:/webkit/WebKitWebView/save (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10183 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2785 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44093 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80415 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2875 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101304 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21300 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15518 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80903 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21552 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81117 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80285 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20017 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24835 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2205 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14507 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21284 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26463 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20971 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24431 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22712 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (approved):
This is delightful! So much less hand-written code!

Now that you noticed this issue, I think we’ll want to add WPT tests that intentionally try different orders where the serialization order should not depend on the order it was specified, to verify that serialization uses a canonical order and not the order passed in. For properties like `hanging-punctuation`.

weinig committed:
Add support for generating a consumer for CSS property that use group terms
https://bugs.webkit.org/show_bug.cgi?id=289305

Reviewed by Darin Adler.

Adds support for generating consumers with "match one or more any order" (eg `[ <foo> || <bar> ]`),
"match all any order" (eg `[ <foo> && <bar> ]`), "match all ordered" (eg `[ <foo> <bar> ]`) and
optional terms (e.g. `<foo>?`).

A new annotation, @(preserves-order) was needed for properties with "any order" grammars (eg || or &&)
but where the order in which things are declared is important such as the `paint-order` property. In
some cases this has been used to maintain existing serialization behavior, such as hanging-punctuation,
but it is not clear that this is necessarily correct and should be evaluated separately.

Together, these allow generation of:
    font-variant-east-asian
    font-variant-ligatures
    font-variant-numeric
    text-autospace
    fill
    stroke
    speak-as
    text-indent
    text-transform
    contain
    hanging-punctuation
    text-decoration-line
    -webkit-text-decorations-in-effect
    text-emphasis-position
    text-emphasis-style
    touch-action
    @font-face font-variant-east-asian
    @font-face font-variant-ligatures
    @font-face font-variant-numeric

Additionally, this change adds generation for the following two properties/descriptors missed
in earlier passes:
    font-width
    @font-face font-weight

While adding generation for font-width, I noticed a few places were still calling
it font-stretch, so I updated those to match the new naming.

* LayoutTests/editing/execCommand/toggle-text-decorations-expected.txt:
* LayoutTests/editing/execCommand/toggle-text-decorations.html:
* LayoutTests/fast/css3-text/css3-text-decoration/getComputedStyle/getComputedStyle-text-decoration-shorthand-ordering-expected.txt:
* LayoutTests/fast/css3-text/css3-text-decoration/getComputedStyle/getComputedStyle-text-decoration-shorthand-ordering.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/parsing/text-decoration-line-valid-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/editing/run/strikethrough_2001-last-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/editing/run/underline_2001-last-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/strikethrough_2001-last-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/editing/run/underline_2001-last-expected.txt:
    - The serialization of text-decorations-line has changed to match the grammar order which is what
      WPT expects.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Contain.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Contain.h: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PointerEvents.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PointerEvents.h: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Speech.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Speech.h: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Text.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Text.h: Removed.
    - Remove no longer needed files.

* Source/WebCore/css/scripts/process-css-properties.py:
    - Replaces single GroupTerm class with MatchAllOrderedTerm, MatchAllAnyOrderTerm
      and MatchOneOrMoreAnyOrderTerm classes.
    - Add support for 'type' and 'no-single-item-opt' annotations to the new term types.
    - Add term generators for the new term types.
    - Add pass-through term generation for OptionalTerm. Other generators currently
      just check if the generators instance matches to implement optional semantics.
    - Add additional comments to generated output with the grammar definitions.

* Source/WebCore/css/CSSProperties.json:
    - Update properties that can be generated to use the grammar.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
    - Call now generated functions instead of hand written ones.

* Source/WebCore/css/CSSFontFaceSet.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+SVG.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+SVG.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Text.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Text.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.h:
    - Remove now generated consume functions.
    - Call now generated functions instead of hand written ones.
    - Update uses of 'font-stretch' naming to use 'font-width' terminology.

* Source/WebCore/css/scripts/test/TestCSSProperties.json:
    - Add test cases for the different grouping types.

* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleDeclaration+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
    - Update test results.

Canonical link: https://commits.webkit.org/291871@main
